File: isabelle.eclipse.core/src/isabelle/eclipse/core/IsabelleCorePlugin.java
package isabelle.eclipse.core;

import isabelle.eclipse.core.app.Isabelle;

import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Plugin;
import org.eclipse.core.runtime.Status;
import org.osgi.framework.BundleContext;

public class IsabelleCorePlugin extends Plugin {

	// The plug-in ID
	public static final String PLUGIN_ID = "isabelle.eclipse.core"; //$NON-NLS-1$

	// The shared instance
	private static IsabelleCorePlugin plugin;
	
	private Isabelle isabelle;
	
	/**
	 * The constructor
	 */
	public IsabelleCorePlugin() {
	}

	/*
	 * (non-Javadoc)
	 * @see org.eclipse.ui.plugin.AbstractUIPlugin#start(org.osgi.framework.BundleContext)
	 */
	@Override
	public void start(BundleContext context) throws Exception {
		super.start(context);
		plugin = this;
		
		isabelle = new Isabelle();
		
	}

	/*
	 * (non-Javadoc)
	 * @see org.eclipse.ui.plugin.AbstractUIPlugin#stop(org.osgi.framework.BundleContext)
	 */
	@Override
	public void stop(BundleContext context) throws Exception {
		
		if (isabelle != null) {
			isabelle.stop();
			isabelle = null;
		}
		
		plugin = null;
		super.stop(context);
	}

	/**
	 * Returns the shared instance
	 *
	 * @return the shared instance
	 */
	public static IsabelleCorePlugin getDefault() {
		return plugin;
	}
	
	public static Isabelle getIsabelle() {
		return getDefault().isabelle;
	}
	
	/**
	 * Writes the message to the plug-in's log
	 * 
	 * @param message
	 *            the text to write to the log
	 * @param exception
	 *            exception to capture in the log
	 */
	public static void log(String message, Throwable exception) {
		IStatus status = error(message, exception);
		getDefault().getLog().log(status);
	}
	
	/**
	 * Logs the exception to the plug-in's log. Exception's message is used as
	 * the log message.
	 * 
	 * @param exception
	 *            exception to capture in the log
	 */
	public static void log(Throwable exception) {
		log(exception.getMessage(), exception);
	}
	
	/**
	 * Returns a new error {@code IStatus} for this plug-in.
	 * 
	 * @param exception
	 *            exception to wrap in the error {@code IStatus}
	 * @return the error {@code IStatus} wrapping the exception
	 */
	public static IStatus error(Throwable exception) {
		return error(exception.getMessage(), exception);
	}
	
	/**
	 * Returns a new error {@code IStatus} for this plug-in.
	 * 
	 * @param message
	 *            text to have as status message
	 * @param exception
	 *            exception to wrap in the error {@code IStatus}
	 * @return the error {@code IStatus} wrapping the exception
	 */
	public static IStatus error(String message, Throwable exception) {
		if (message == null) {
			message = ""; 
		}
		return new Status(IStatus.ERROR, PLUGIN_ID, 0, message, exception);
	}
	
}

File: isabelle.eclipse.core/src/isabelle/eclipse/core/internal/IsabelleCorePlugin.scala
package isabelle.eclipse.core.internal

import org.eclipse.core.runtime.{IStatus, Plugin, Status}
import org.osgi.framework.BundleContext

import isabelle.eclipse.core.app.Isabelle


/**
 * @author Andrius Velykis
 */
object IsabelleCorePlugin {

  // The shared instance
  private var instance: IsabelleCorePlugin = _
  def plugin = instance

  def log(status: IStatus) = plugin.getLog.log(status)

  /**
   * Returns a new error `IStatus` for this plug-in.
   *
   * @param message    text to have as status message
   * @param exception  exception to wrap in the error `IStatus`
   * @return  the error `IStatus` wrapping the exception
   */
  def error(ex: Option[Throwable] = None, msg: Option[String] = None): IStatus = {

    // if message is not given, try to use exception's
    val message = msg orElse { ex.flatMap(ex => Option(ex.getMessage)) }

    new Status(IStatus.ERROR, plugin.pluginId, 0, message.orNull, ex.orNull);
  }
}


/**
 * @author Andrius Velykis
 */
class IsabelleCorePlugin extends Plugin {

  IsabelleCorePlugin.instance = this

  // The plug-in ID
  def pluginId = "isabelle.eclipse.core" //$NON-NLS-1$


  var isabelleInit = false

  /**
   * The Isabelle prover instance (may be not started)
   */
  lazy val isabelle: Isabelle = {
    isabelleInit = true
    new Isabelle
  }

  override def stop(context: BundleContext) {

    if (isabelleInit) {
      isabelle.stop()
    }

    super.stop(context)
  }

}
